### PR TITLE
Feature/versioned inheritance

### PR DESCRIPTION
**Pull request by bulbozaur**

### CODE REVIEW

File: contracts/0.4.24/utils/Versioned.sol [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// SPDX-FileCopyrightText: 2022 Lido <info@lido.fi>
```
Comment (TheDZhon) [inline contracts/0.4.24/utils/Versioned.sol:1]: 2023

File: contracts/0.4.24/utils/Versioned.sol [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    /// Version of the initialized contract storage.
+    /// The version stored in CONTRACT_VERSION_POSITION equals to:
+    /// - 0 right after the deployment, before an initializer is invoked (and only at that moment);
+    /// - N after calling initialize(), where N is the initially deployed contract version;
+    /// - N after upgrading contract by calling finalizeUpgrade_vN().
+    bytes32 internal constant CONTRACT_VERSION_POSITION = keccak256("lido.Versioned.contractVersion");
+
+    constructor() {
+        // lock version in the implementation's storage to prevent initialization
+        CONTRACT_VERSION_POSITION.setStorageUint256(2**256 - 1);
```
Comment (TheDZhon) [inline contracts/0.4.24/utils/Versioned.sol:23]: maybe internal constant
```
uint256 internal constant PERTIFIED_VERSION_MARK = uint256(-1); // or 2**256-1 as you wish
```

File: contracts/0.4.24/utils/Versioned.sol [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    /// @notice Returns the current contract version.
+    function getContractVersion() public view returns (uint256) {
```
Comment (TheDZhon) [inline contracts/0.4.24/utils/Versioned.sol:27]: maybe `external`?

Comment (bulbozaur) [reply contracts/0.4.24/utils/Versioned.sol:29]: Its used in _checkContractVersion and only while upgrade/init, but extra internal method will add extra byte code to Lido and NOR contracts, maybe inline getStorageUint256 in _checkContractVersion and make getContractVersion 
 external is a way?

File: contracts/0.4.24/oracle/LidoOracle.sol [hunk 4]
```diff
@@ -30,7 +32,7 @@ import "./ReportUtils.sol";
  * Not all frames may come to a quorum. Oracles may report only to the first epoch of the frame and
  * only if no quorum is reached for this epoch yet.
  */
-contract LidoOracle is AragonApp {
+contract LidoOracle is AragonApp, Versioned {
```
Comment (skozin) [inline contracts/0.4.24/oracle/LidoOracle.sol:35]: This contract is being re-written from scratch with the absolute most of the code removed.

File: contracts/0.4.24/utils/Versioned.sol (hunk 3 above)
Comment (TheDZhon) [reply contracts/0.4.24/utils/Versioned.sol:29]: I see, you are right

File: contracts/0.4.24/utils/Versioned.sol [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    /// @dev Storage slot: uint256 version
+    /// Version of the initialized contract storage.
+    /// The version stored in CONTRACT_VERSION_POSITION equals to:
+    /// - 0 right after the deployment, before an initializer is invoked (and only at that moment);
+    /// - N after calling initialize(), where N is the initially deployed contract version;
+    /// - N after upgrading contract by calling finalizeUpgrade_vN().
+    bytes32 internal constant CONTRACT_VERSION_POSITION = keccak256("lido.Versioned.contractVersion");
+
+    uint256 internal constant PERTIFIED_VERSION_MARK = uint256(-1);
```
Comment (arwer13) [inline contracts/0.4.24/utils/Versioned.sol:21]: Let's refactor 0.9.8 version in accordance?
- add `PERTIFIED_VERSION_MARK`
- replace `getContractVersion() external` and `_getContractVersion()` by `getContractVersion() public`

- I'd also like to refactor this in 0.8.9 version:
```
        if (_getContractVersion() == 0) {
            _writeContractVersion(1);
        } else {
            revert NonZeroContractVersionOnInit();
        }
```
to the common "require-like" style:
```
        if (_getContractVersion() != 0) {
            revert NonZeroContractVersionOnInit();
        }
        _writeContractVersion(1);
```

File: contracts/0.4.24/utils/Versioned.sol [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    /// @notice Returns the current contract version.
+    function getContractVersion() public view returns (uint256) {
+        return CONTRACT_VERSION_POSITION.getStorageUint256();
+    }
+
+    function _checkContractVersion(uint256 version) internal view {
+        require(version == getContractVersion(), "UNEXPECTED_CONTRACT_VERSION");
+    }
+
+    function _writeContractVersion(uint256 version) internal {
```
Comment (arwer13) [inline contracts/0.4.24/utils/Versioned.sol:37]: Let's add a comment that logic related to `InvalidContractVersionIncrement` is not added here to save contract size for 0.4.24 overweighed contracts.

**Review by arwer13 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**